src/display3d.cpp

```diff
@@ -1717,26 +1702,18 @@ void displayBlueprints(const glm::mat4 &viewMatrix)
 			state = SS_BLUEPRINT_INVALID;
 		}
 		// we are placing a building or a delivery point
-		if (sBuildDetails.psStats->ref >= REF_STRUCTURE_START
-		    && sBuildDetails.psStats->ref < (REF_STRUCTURE_START + REF_RANGE))
+		if (STRUCTURE_STATS *stats = castStructureStats(sBuildDetails.psStats))
 		{
 			// it's a building
+			uint16_t direction = snapDirection(player.r.y);
 			if (wallDrag.status == DRAG_PLACING || wallDrag.status == DRAG_DRAGGING)
 			{
-				int left, right, up, down;
-				// a wall (or something like that)
-
-				left = MIN(wallDrag.x1, wallDrag.x2);
-				right = MAX(wallDrag.x1, wallDrag.x2);
-				up = MIN(wallDrag.y1, wallDrag.y2);
-				down = MAX(wallDrag.y1, wallDrag.y2);
-
-				drawLineBuild((STRUCTURE_STATS *)sBuildDetails.psStats, left, right, up, down, (player.r.y + 0x2000) & 0xC000, state);
+				drawLineBuild(stats, wallDrag.pos, wallDrag.pos2, direction, state);
 			}
 			else
 			{
 				unsigned width, height;
-				if (((player.r.y + 0x2000) & 0x4000) == 0)
+				if ((direction & 0x4000) == 0)
```

Comment:
A few of these were essentially changed to

`((player.r.y + 0x2000) & 0xC000) & 0x4000`

Is that going to be a problem?

Reply:
I think it should be equivalent, since `((player.r.y + 0x2000) & 0xC000) & 0x4000` = `(player.r.y + 0x2000) & (0xC000 & 0x4000)` = `(player.r.y + 0x2000) & 0x4000`.